api/src/main/scala/org/ensime/api/common.scala

```diff
@@ -129,10 +131,80 @@ object OffsetRange extends ((Int, Int) => OffsetRange) {
 // such as files/dirs, existance, content hints
 // (java/scala/class/resource) in the type, validated at construction
 // (and can be revalidated at any time)
-sealed trait EnsimeFile
-final case class RawFile(file: Path) extends EnsimeFile
+sealed trait EnsimeFile {
+  def path: Path
```

Comment:
you cannot relicense GPLv3 code under Apache 2.0, it would change the license of the `api` module and then our downstream developers would be very annoyed because they'd have to relicense their applications or fork ensime-server.

If you want to know about the benefits of separating data and functionality, read chapter 4 of my book. In this particular case, consider an Emacs, Atom, VSCode or Vim developer looking at our API, why would they care about 100 lines of distracting Scala code when they just want to see the data structure to be able to interpret the wire protocol format? Furthermore, within ensime itself if we move to `EnsimeFile` there is a **lot** of data conveyed when a method accepts an `EnsimeFile`, `ArchiveFile` or `RawFile`. Indeed, the problem with `FileObject` is that it tries to do exactly what you describe (have lots of methods) but what we really need is the contextual understanding of *what* we have in our hands.

Another example are the data structures we use for storage in the indexer. The actual implementation, which uses large OOP classes, are so complicated to explain that the only way I could describe them for the Scala Sphere talk was to rewrite them as pure data structures and type classes.

Basically splitting data structure and functionality leads to cleaner modular code that is easier to reason about.